<a href="https://colab.research.google.com/github/TiagoHarari/DataAnalysis_course/blob/main/1-indexing_ejercicios.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<a href="https://colab.research.google.com/github/institutohumai/cursos-python/blob/master/AnalisisDeDatos/1_Indexing/ejercicio/ejercicio.ipynb"> <img src='https://colab.research.google.com/assets/colab-badge.svg' /> </a>
<div align="center"> Recordá abrir en una nueva pestaña </div>


Para este ejercicio vamos a importar un conjunto de datos con el listado de colegios públicos de Bogotá del año 2016

# Ejercicio I: base de Colegios de Bogotá.

In [1]:
# primero importamos las librerias a utilizar
import pandas as pd
import numpy as np

In [2]:
#definimos el dataframe de nuestro archivo CSV a utilizar
df_colegios = pd.read_csv('https://www.datos.gov.co/api/views/qijw-htwa/rows.csv?accessType=DOWNLOAD')

¿Cuáles son las columnas? ¿Cuántas filas tiene?

In [ ]:
print("cantidad de columnas: ", df_colegios.columns)

cantidad de columnas:  Index(['año', 'secretaria', 'nombreestablecimiento', 'zona', 'direccion',
       'telefono', 'nombre_Rector', 'tipo_Establecimiento', 'etnias', 'sector',
       'genero', 'niveles', 'jornadas', 'caracter', 'especialidad', 'grados',
       'modelos_Educativos', 'capacidades_Excepcionales', 'discapacidades',
       'idiomas', 'numero_de_Sedes', 'estado', 'prestador_de_Servicio',
       'propiedad_Planta_Fisica', 'resguardo', 'matricula_Contratada',
       'calendario', 'internado', 'estrato_Socio_Economico',
       'correo_Electronico'],
      dtype='object')


In [ ]:
df_colegios.shape

(2404, 30)

In [ ]:
# para intorudcir a la visualizacion grafica
df_colegios.head()

año secretaria                              nombreestablecimiento    zona  \
0  2016     BOGOTA  INST DE BTO TEC COMERCIAL PITAGORAS           ...  URBANA   
1  2016     BOGOTA                 GIMNASIO INFANTIL FEDERICO FROEBEL  URBANA   
2  2016     BOGOTA             GIMNASIO INFANTIL LOS NIÑOS DEL MAÑANA  URBANA   
3  2016     BOGOTA               COLEGIO FRANCISCO PRIMERO S.S. (IED)  URBANA   
4  2016     BOGOTA  COL SAN JUAN VIANETH                          ...  URBANA   

           direccion             telefono                     nombre_Rector  \
0         KR 5 11 67  5665677   / 3429092                YOLANDA ARIAS CRUZ   
1  CL 32 SUR 52 A 54              7100349              SANDRA MILENA MENDEZ   
2     CL 134 D 47 27              6010350           CLAUDIA BOCACHICA SAENZ   
3        KR 28 63 64   2126019 -- 2481271      JOSE CUSTODIO GALARZA MORENO   
4    KR 2 C 3 18 SUR   2804738 -- 4714934  GLORIA ESPERANZA CHUQUIN LAISECA   

    tipo_Establecimiento etnias  sector  ...  numero_de_Sedes estado  \
0  INSTITUCION EDUCATIVA    NaN     NaN  ...                1    NaN   
1       CENTRO EDUCATIVO    NaN     NaN  ...                1    NaN   
2  INSTITUCION EDUCATIVA    NaN     NaN  ...                1    NaN   
3  INSTITUCION EDUCATIVA    NaN     NaN  ...                4    NaN   
4       CENTRO EDUCATIVO    NaN     NaN  ...                1    NaN   

  prestador_de_Servicio  propiedad_Planta_Fisica  resguardo  \
0       PERSONA NATURAL          PERSONA NATURAL  NO APLICA   
1       PERSONA NATURAL          PERSONA NATURAL  NO APLICA   
2       PERSONA NATURAL          PERSONA NATURAL  NO APLICA   
3               OFICIAL                  OFICIAL  NO APLICA   
4       PERSONA NATURAL          PERSONA NATURAL  NO APLICA   

  matricula_Contratada calendario internado estrato_Socio_Economico  \
0                   NO          A       NaN                     NaN   
1                   NO          A       NaN                     NaN   
2                   NO          A       NaN                     NaN   
3                   NO          A       NaN                     NaN   
4                   NO          A       NaN                     NaN   

                                  correo_Electronico  
0                         pitagoras@pitagoras.edu.co  
1                             gimfedebel@hotmail.com  
2                claudialindenibocachica@hotmail.com  
3                          cedalemania12@redp.edu.co  
4  colegiosanjuanviateh@hotmail.com -- colegiosan...  

[5 rows x 30 columns]

In [ ]:
df_colegios.sample(5)

año secretaria                              nombreestablecimiento  \
2283  2016     BOGOTA                      COLEGIO PALERMO  IEDIP  (IED)   
214   2016     BOGOTA                         CENT EDUC DIST DON BOSCO V   
1161  2016     BOGOTA  COL DEL SANTO ANGEL                           ...   
1725  2016     BOGOTA  JARDIN INFANTIL Y PREESCOLAR PEQUEÑOS TALENTOS...   
1735  2016     BOGOTA  COL KAPEIROT                                  ...   

        zona         direccion  \
2283  URBANA       KR 23 49 37   
214   URBANA   CL 143 A 118 50   
1161  URBANA       CL 165 7 21   
1725  URBANA      CL 122 20 60   
1735  URBANA  CL 8 SUR 38 B 05   

                                               telefono  \
2283                      3203220 -- 2850618 -- 2857194   
214                                             6900123   
1161        6711524 -- 6711087 -- 6784677 -- 3115182794   
1725                                    6130165-6203100   
1735  2025581 / 2038793                             ...   

                      nombre_Rector   tipo_Establecimiento            etnias  \
2283          MARBEL CALA RODRIGUEZ  INSTITUCION EDUCATIVA  AFRODESCENDIENTE   
214      ROSEMBEL SUAREZ BETANCOURT  INSTITUCION EDUCATIVA               NaN   
1161  HNA ISABEL GARCIA DE LA PARTE  INSTITUCION EDUCATIVA               NaN   
1725        CLAUDIA YANET RODRIGUEZ       CENTRO EDUCATIVO               NaN   
1735       GERARDO MURCIA RODRIGUEZ  INSTITUCION EDUCATIVA               NaN   

      sector  ...  numero_de_Sedes estado prestador_de_Servicio  \
2283     NaN  ...                1    NaN               OFICIAL   
214      NaN  ...                1    NaN               OFICIAL   
1161     NaN  ...                1    NaN   COMUNIDAD RELIGIOSA   
1725     NaN  ...                1    NaN       PERSONA NATURAL   
1735     NaN  ...                1    NaN       PERSONA NATURAL   

      propiedad_Planta_Fisica  resguardo matricula_Contratada calendario  \
2283                  OFICIAL  NO APLICA                   NO          A   
214                   OFICIAL  NO APLICA                   SI          A   
1161      COMUNIDAD RELIGIOSA  NO APLICA                   NO          A   
1725          PERSONA NATURAL  NO APLICA                   NO          B   
1735          PERSONA NATURAL  NO APLICA                   NO          A   

     internado estrato_Socio_Economico  \
2283       NaN                     NaN   
214        NaN                     NaN   
1161       NaN                     NaN   
1725       NaN                     NaN   
1735       NaN                     NaN   

                                     correo_Electronico  
2283                      cedpalermocedip13@redp.edu.co  
214   johnalbertr36@gmail.com -- gestionacademicabos...  
1161   secretaria.academica@colegiodelsantoangel.edu.co  
1725                     crodriguez56@littletalents.com  
1735  colegiokapeirot@yahoo.com  -  kapeirotadmision...  

[5 rows x 30 columns]

Seleccionen los colegios donde la propiedad de la planta física sea de una persona natural o comunidad religiosa.

In [3]:
df_colegios["propiedad_Planta_Fisica"].value_counts()
# de la columna: propiedad_Planta_Fisica contamos sus valores con value_counts

propiedad_Planta_Fisica
PERSONA NATURAL              1727
OFICIAL                       405
COMUNIDAD RELIGIOSA           151
FUNDACION O CORPORACIONES      46
FEDERACIONES                   17
REGIMEN ESPECIAL               16
COOPERATIVO                    14
COMUNIDAD                      13
CAJA DE COMPENSACIÓN            9
SOCIEDAD                        6
Name: count, dtype: int64

In [6]:
propiedad = (df_colegios['propiedad_Planta_Fisica'] == "PERSONA NATURAL") | (df_colegios['propiedad_Planta_Fisica'] == "COMUNIDAD RELIGIOSA")
df_colegios[propiedad].tail()

año secretaria                              nombreestablecimiento  \
2399  2016     BOGOTA                       COLEGIO SAN JOSE DE CALASANZ   
2400  2016     BOGOTA                        COL NUEVO REINO DE TURINGIA   
2401  2016     BOGOTA  LIC  INF GOOFY                                ...   
2402  2016     BOGOTA  LIC APRENDO CON MIS AMIGOS                    ...   
2403  2016     BOGOTA  JARD INF EL ROSAL DE MI INFANCIA              ...   

        zona             direccion                              telefono  \
2399  URBANA        CL 129 88 B 35        6854492 -- 6854495 -- 6803549    
2400  URBANA     KR 103 B 152 C 34                               6830915   
2401  URBANA   CL 65 F SUR 78 C 32  7761739 / 7762538 / 7753157 /8045802   
2402  URBANA       CR. 94 NO.29-30                               2671099   
2403  URBANA  TR. 56A NO.45-18 SUR                               7242172   

                          nombre_Rector   tipo_Establecimiento etnias  sector  \
2399              GLADIS CUÉLLAR IBAÑEZ  INSTITUCION EDUCATIVA    NaN     NaN   
2400        RUTH CECILIA CORREA FAJARDO       CENTRO EDUCATIVO    NaN     NaN   
2401  SANDRA PATRICIA SANTAMARIA OVALLE       CENTRO EDUCATIVO    NaN     NaN   
2402    YANNETH CONSUELO BOLIVAR GAITAN       CENTRO EDUCATIVO    NaN     NaN   
2403                    ADALGIZA LADINO       CENTRO EDUCATIVO    NaN     NaN   

      ...  numero_de_Sedes estado prestador_de_Servicio  \
2399  ...                1    NaN   COMUNIDAD RELIGIOSA   
2400  ...                1    NaN       PERSONA NATURAL   
2401  ...                1    NaN       PERSONA NATURAL   
2402  ...                1    NaN       PERSONA NATURAL   
2403  ...                1    NaN       PERSONA NATURAL   

      propiedad_Planta_Fisica  resguardo matricula_Contratada calendario  \
2399      COMUNIDAD RELIGIOSA  NO APLICA                   NO          A   
2400          PERSONA NATURAL  NO APLICA                   NO          A   
2401          PERSONA NATURAL  NO APLICA                   NO          A   
2402          PERSONA NATURAL  NO APLICA                   NO          A   
2403          PERSONA NATURAL  NO APLICA                   NO          A   

     internado estrato_Socio_Economico  \
2399       NaN                     NaN   
2400       NaN                     NaN   
2401       NaN                     NaN   
2402       NaN                     NaN   
2403       NaN                     NaN   

                                     correo_Electronico  
2399                                  csjcsuba@yahoo.es  
2400  jibnuevoreinodeturingia@hotmail.com - ruth_cec...  
2401  pattysantaco@yahoo.com   directoragoofy@gmail....  
2402                                                NaN  
2403  adalgizaladino@hotmail.com                    ...  

[5 rows x 30 columns]

¿Hay alguno de estos colegios que sea para capacidades excepcionales?



```
# El método unique() devuelve los valores únicos de una Serie. Es útil para ver rápidamente los distintos valores que contiene una columna.

```



In [7]:
# analizamos los distintos valores que contiene la columna capacidades excepcionales
df_colegios['capacidades_Excepcionales'].unique()

array([nan, 'CAPACIDADES EXCEPCIONALES',
       'TALENTO CIENTIFICO,TALENTO TECNOLOGICO,TALENTO SUBJETIVO/ARTÍSTICO',
       'TALENTO SUBJETIVO/ARTÍSTICO', 'TALENTO CIENTIFICO'], dtype=object)

¿Cuáles son los colegios que ofrecen todos los niveles educativos? ¿Qué idiomas ofrecen?

In [ ]:
df_colegios[['nombreestablecimiento','niveles']].sample(5)

nombreestablecimiento  \
167   JARD PSICOPEDAG EL MANANTIAL DEL SABER        ...   
1870            COLEGIO TÉCNICO  JAIME PARDO LEAL (IED)   
2014  JARD INF PYLOSOS                              ...   
1601  CAMPANITA BILINGUAL PRESCHOOL                 ...   
1330  COL MERCEDARIO SAN PEDRO NOLASCO              ...   

                                                niveles  
167                          PREESCOLAR,BÁSICA PRIMARIA  
1870  PREESCOLAR,MEDIA,BÁSICA SECUNDARIA,BÁSICA PRIM...  
2014                                         PREESCOLAR  
1601                                         PREESCOLAR  
1330  PREESCOLAR,MEDIA,BÁSICA SECUNDARIA,BÁSICA PRIM...

In [ ]:
df_colegios["niveles"].value_counts()
# para todos los niveles, hago una serie con todos los valores contados

niveles
PREESCOLAR,MEDIA,BÁSICA SECUNDARIA,BÁSICA PRIMARIA                     1068
PREESCOLAR,BÁSICA PRIMARIA                                              491
PREESCOLAR                                                              456
MEDIA,BÁSICA SECUNDARIA,BÁSICA PRIMARIA                                 135
PREESCOLAR,BÁSICA SECUNDARIA,BÁSICA PRIMARIA                             95
MEDIA,BÁSICA SECUNDARIA                                                  81
BÁSICA PRIMARIA                                                          26
PREESCOLAR,BÁSICA SECUNDARIA,MEDIA,BÁSICA PRIMARIA                        9
MEDIA                                                                     6
PREESCOLAR,MEDIA,BÁSICA SECUNDARIA,BÁSICA PRIMARIA,PRIMERA INFANCIA       5
PREESCOLAR,MEDIA,BÁSICA PRIMARIA                                          4
PREESCOLAR,MEDIA,BÁSICA SECUNDARIA                                        3
PREESCOLAR,BÁSICA SECUNDARIA                                              3
BÁSI

In [9]:
# 'PREESCOLAR,MEDIA,BÁSICA SECUNDARIA,BÁSICA PRIMARIA,PRIMERA INFANCIA' --> son todos los niveles posibles de un colegio
df_niveles = df_colegios[df_colegios['niveles'] == 'PREESCOLAR,MEDIA,BÁSICA SECUNDARIA,BÁSICA PRIMARIA,PRIMERA INFANCIA']


In [10]:
# ¿que idiomas ofrecen?
idiomas = df_niveles[df_niveles['idiomas'].notna()]['idiomas'].unique()
# booleana donde cada valor indica si el correspondiente valor en la columna idiomas no es nulo (True si no es nulo, False si es nulo).
# solo con valores que no son nan (unique)
print("los idiomas que dictan son: ", idiomas)

los idiomas que dictan son:  ['INGLÉS']


# Ejercicio II: PyEPH

### PyEPH

La librería Pyeph tiene como objetivo facilitar el procesamiento en Python de las Encuesta Permanente de Hogares (eph) publicadas por INDEC de forma periódica. Está pensada como un espacio donde se nuclean y centralizan los cálculos vinculados a las mismas para posteriormente ser utilizadas en investigaciones, artículos, publicaciones, etc. Es una librería que hace principal hincapié en la transparencia metodológica utilizando licencias de código abierto y que promueve la colaboración de las comunidades de cientístas de datos, sociales, investigadorxs, desarrolladorxs, periodistas y demás curiosxs.

https://github.com/institutohumai/pyeph

Para entender qué representan las columnas ver documentación:


https://www.indec.gob.ar/ftp/cuadros/menusuperior/eph/EPH_diseno_reg_t414.pdf

In [1]:
# Instalamos la librería.

!pip install pyeph

In [2]:
import pyeph

In [3]:
# Ejemplo: traemos el 2do trimestre del 2021
eph = pyeph.get(data="eph", year=2021, period=2, base_type='individual') # EPH individual


Obtenido con exito: base_individual_2021T2.zip 


In [4]:
eph.columns

Index(['CODUSU', 'ANO4', 'TRIMESTRE', 'NRO_HOGAR', 'COMPONENTE', 'H15',
       'REGION', 'MAS_500', 'AGLOMERADO', 'PONDERA',
       ...
       'PDECIFR', 'ADECIFR', 'IPCF', 'DECCFR', 'IDECCFR', 'RDECCFR', 'GDECCFR',
       'PDECCFR', 'ADECCFR', 'PONDIH'],
      dtype='object', length=177)

In [5]:
eph.head()

CODUSU  ANO4  TRIMESTRE  NRO_HOGAR  COMPONENTE  H15  \
0  TQRMNOPYQHMLKNCDEHJGH00699689  2021          2          1           1    1   
1  TQRMNOPYQHMLKNCDEHJGH00699689  2021          2          1           2    1   
2  TQRMNOPYSHMLKNCDEHJGH00699098  2021          2          1           1    1   
3  TQRMNOPYSHMLKNCDEHJGH00699098  2021          2          1           2    1   
4  TQRMNOPYSHMLKNCDEHJGH00699098  2021          2          1           3    1   

   REGION MAS_500  AGLOMERADO  PONDERA  ...  PDECIFR  ADECIFR   IPCF  DECCFR  \
0      40       S          23      214  ...      NaN        5  27000       6   
1      40       S          23      214  ...      NaN        5  27000       6   
2      40       S          23      192  ...      NaN       10  46000       9   
3      40       S          23      192  ...      NaN       10  46000       9   
4      40       S          23      192  ...      NaN       10  46000       9   

   IDECCFR  RDECCFR  GDECCFR  PDECCFR  ADECCFR  PONDIH  
0      7.0        8      6.0      NaN        8     218  
1      7.0        8      6.0      NaN        8     218  
2      9.0       10      8.0      NaN        9     196  
3      9.0       10      8.0      NaN        9     196  
4      9.0       10      8.0      NaN        9     196  

[5 rows x 177 columns]

**Primer paso.** Crear una variable llamada eph_2 que contenga las siguientes columnas para realizar el siguiente análisis:

Edad (**CH_06**), Sexo (**CH_04**), Ingreso de la ocupación principal (**P21**), Categoría ocupacional (**CAT_OCUP**), Condición de actividad (ESTADO), Ponderación (**PONDERA**)

Referencias:

Sexo:
*   01 = Varón
*   02 = Mujer

Categoría ocupacional (para ocupados y desocupados con
ocupación anterior):
*   1 = Patrón
*   2 = Cuenta propia
*   3 = Obrero o empleado
*   4 = Trabajador familiar sin remuneración
*   9 = Ns/Nr

Estado - Condición de actividad:
*   0 = Entrevista individual no realizada
*   1 = Ocupado
*   2 = Descoupado
*   3 = Inactivo
*   4 = menor de 10 años

In [6]:
#Filtramos por columnas
eph_2=eph[["CH06", "CH04", "P21", "CAT_OCUP", "ESTADO", "PONDERA"]]

In [7]:
#Renombramos columnas
eph_2.columns=["edad", "sexo", "ingreso", "cat_ocup", "estado", "pondera"]

La columna **pondera** representa un estimado poblacional de las personas que representa cada fila. Por lo tanto, para cada cálculo es necesario operar con esa columna. Por ejemplo, la fila 0 representa a 214 personas de 58 años de sexo masculino, que tienen $50.000 de ingreso de la ocupación principal, su categoría es cuentapropista y pertenecen al segmento de ocupados.   

In [8]:
eph_2.head()

edad  sexo  ingreso  cat_ocup  estado  pondera
0    58     1    50000         2       1      214
1    24     1        0         3       2      214
2    51     2        0         4       1      192
3    27     1     4000         2       1      192
4    18     1        0         0       3      192

In [9]:
# analizar tamaño del dataframe
eph_2.shape

(47085, 6)

*214 personas (pondera) de 58 años (edad) de sexo masculino (sexo), que tienen $50.000 de ingreso (ingreso) de la ocupación principal (cat_ocup), su categoría es cuentapropista y pertenecen al segmento de ocupados (estado).*

a) Traer un dataframe que filtre solo a las personas que tengan 21 años o menos.

In [14]:
eph_2[eph_2['edad'] <= 21]

edad  sexo  ingreso  cat_ocup  estado  pondera
4        18     1        0         0       3      192
8         5     1        0         0       4      250
11       19     2        0         0       3      250
12       17     1        0         0       3      250
13       14     1        0         0       3      250
...     ...   ...      ...       ...     ...      ...
47070    15     2        0         0       3       46
47071    13     1        0         0       3       46
47074     8     1        0         0       4       49
47077    10     1        0         0       3      111
47084    20     1     4000         2       1      230

[15192 rows x 6 columns]

In [13]:
menores_de_21 = eph_2[eph_2['edad'] <= 21]
print(menores_de_21)

       edad  sexo  ingreso  cat_ocup  estado  pondera
4        18     1        0         0       3      192
8         5     1        0         0       4      250
11       19     2        0         0       3      250
12       17     1        0         0       3      250
13       14     1        0         0       3      250
...     ...   ...      ...       ...     ...      ...
47070    15     2        0         0       3       46
47071    13     1        0         0       3       46
47074     8     1        0         0       4       49
47077    10     1        0         0       3      111
47084    20     1     4000         2       1      230

[15192 rows x 6 columns]


***hay 15.192 personas menores a 21 años dentro de la muestra***

b) Traer un dataframe que filtre a las personas desocupadas de más de 40 años.

In [18]:
# estado = 2 --> desocupado
eph_2[(eph_2['estado'] == 2) & (eph_2['edad'] > 40)]

edad  sexo  ingreso  cat_ocup  estado  pondera
47       58     2        0         3       2      915
77       41     1        0         2       2      139
92       45     1        0         3       2      127
357      42     1        0         3       2     1086
433      48     2        0         3       2      626
...     ...   ...      ...       ...     ...      ...
46276    54     1        0         2       2      593
46350    49     1        0         3       2      307
46447    72     1        0         2       2      232
46599    49     2        0         3       2      311
46715    49     2        0         3       2      215

[503 rows x 6 columns]

***Segun este informe de INDEC, existen 503 personas mayores de 40 años en estado de desocupacion (no ponderado)***

c) ¿Cuántas mujeres hay en la base de datos? ¿Y hombres?

In [19]:
# sexo --> varon = 1 y mujer = 2
print(eph_2['sexo'].value_counts())

2    24483
1    22602
Name: sexo, dtype: int64


In [20]:
eph_mujeres = eph_2[eph_2['sexo'] == 2]
eph_hombres = eph_2[eph_2['sexo'] == 1]

print('El total de mujeres es:', eph_mujeres['pondera'].sum())
print("El total de hombres es: ", eph_hombres['pondera'].sum())

El total de mujeres es: 14833824
El total de hombres es:  14038615


***hay 14.833.824 mujeres y 14.038.615 hombres***

d) ¿Cuál era la tasa de desocupación en el período seleccionado para las mujeres? ¿Era más alta o más baja que la de los hombres? (Extra: pensar el resultado usando una función)

Nota: la tasa se calcula como Desocupados / (Ocupados + Desocupados)

In [24]:
# tasa de desocupacion = Desocupados / (Ocupados + Desocupados)
def obtener_tasa_desocupacion(df): # definimos una funcion
  eph_ocupados = df[df['estado'] == 1] # df de personas en estado de ocupacion
  eph_desocupados = df[df['estado'] == 2] #df de personas en estado de desocupacion
  total_ocupados = eph_ocupados['pondera'].sum() #total ponderado de ocupados
  total_desocupados = eph_desocupados['pondera'].sum() #total ponderado de desocupados

  return round(total_desocupados / (total_ocupados + total_desocupados)*100, 3) # tasa de desocupacion redondeada a 3 decimales

tasa_mujer = obtener_tasa_desocupacion(eph_mujeres)
tasa_hombre = obtener_tasa_desocupacion(eph_hombres)
print("La tasa de desocupacion de las mujeres es: ",tasa_mujer)
print("La tasa de desocupacion de los hombres es: ",tasa_hombre)

if tasa_mujer > tasa_hombre:
  print("La tasa de desocupacion de mujeres es mayor que la de hombres")
else:
  print("La tasa de desocupacion de los hombres es mayor que la de mujeres")


La tasa de desocupacion de las mujeres es:  10.387
La tasa de desocupacion de los hombres es:  9.006
La tasa de desocupacion de mujeres es mayor que la de hombres
